### PR TITLE
LFNext infrastructure and change password migration

### LINKED ISSUE #1309: Establish infrastructure for LFNext app
work occurring in https://github.com/sillsdev/web-languageforge/tree/next/changepassword

### PR DESCRIPTION
**Pull request by longrunningprocess**
## Description

This PR introduces a brand new infrastructure and architecture to the LF app.  It enables a capability by capability migration from the existing Angularjs impl to a modern stack in Svelte and Sveltekit.

Fixes #1309 

## Type of Change

- New feature (non-breaking change which adds functionality)

## Testing on your branch

- [ ] Change password
1. Log in to the app
1. Choose to change your password
1. Walk through the change password pages
1. Attempt to log back in with a new, different password

- [ ] Register for an account
- [ ] Create a non-S/R project
- [ ] Work with a S/R project
- [ ] Attach an image to an entry (< 60 MB)
- [ ] Attach an audio file to an entry (< 60 MB)
- [ ] Attach a file to an entry (> 60 MB)

## TODO In This PR

- [x] Turn off dark mode
- [x] Deploy out to staging (manually)
- [x] Build out automated staging deployment
- [x] finalize the `TODO:`'s in `staging.yml`
- [x] Determine deployment strategy
- [x] Build out automated production deployment
- [ ] think about whether these adhoc builder actions are needed, e.g., `build-and-publish-next.yml`
- [x] Need a way “back” into the legacy app from the next app
- [x] Test the mobile UI ([screenshots below](https://github.com/sillsdev/web-languageforge/pull/1361#issuecomment-1083261967)).
- [ ] Fix up change password tests to suit
- [x] Use self-signed SSL cert in proxy (local dev)
- [x] update config, move @apply out, move class into elements
- [x] turn e2e tests back on in `integrate-and-deploy.yml`?
- [x] verify whether autoprefixer and the postcss.config are still needed or not.

## Production deployment strategy

1. take note of current images in prod
1. merge to `master`
1. set kube context to `prod` from local machine
1. `make delete` from `deployment` dir
1. cut the release
1. monitor the "Deploy to production" action

## For future PR's

- [ ] Implement the user-logged-in menu
- [ ] mock up new header design and navigation
- [ ] require a strong password?
- [ ] consolidate `make dev` and `make next-dev`

## Checklist

- [x] I have performed a self-review of my own code
- [x] I have reviewed the title/description of this PR which will be used as the squashed PR commit message
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have added tests that prove my fix is effective or that my feature works

## qa.languageforge.org testing

Reviewers: add/replace your name below and check the box to sign-off/attest the feature works as expected on qa.languageforge.org

- [ ] Reviewer1 (YYYY-MM-DD HH:MM)
- [ ] Reviewer2 (YYYY-MM-DD HH:MM)

### CODE REVIEW

File: .github/workflows/integrate-and-deploy.yml
Comment (megahirt) [inline .github/workflows/integrate-and-deploy.yml:100]: I wasn't able to delete the `outputs:` line in my suggestion :( 
```suggestion
```

I'm going on a hunch here that these variables are already in the job output, as defined above.  If they are not, then, I suppose this would simply need to be changed to e.g.
`IMAGE_APP: ${{ steps.image.outputs.IMAGE_APP}}`
It would be nice if that wasn't necessary, but I don't actually know

File: docker/next-app/Dockerfile.next-app
Comment (megahirt) [inline docker/next-app/Dockerfile.next-app:10]: after this PR is merged, we can move npm install out of docker

File: next-app/.npmrc
Comment (megahirt) [inline next-app/.npmrc:1]: nice!  We should also put this in the root folder to prevent the same sort of issues with wrong node version

**Review by megahirt — APPROVED** [review]
Wow, a HUGE PR with a lot of groundwork laid.  Thanks for your work on this - excited to see Svelte and SvelteKit land in the LF project!  Just one non-blocking suggestion about refactoring variable names in the GA yaml, but I haven't tested it so I don't know if it works.

**Comment by megahirt** [comment]
@rmunn and I were having a good laugh at the "good first issue" label you put on here, @longrunningprocess :)

File: .github/workflows/integrate-and-deploy.yml
Comment (longrunningprocess) [reply .github/workflows/integrate-and-deploy.yml:53]: yeah, I like this.  I appreciate you trying to do it for me in the suggestion but your namespace is in the wrong place so I'll just push a commit with these changes.

File: .github/workflows/integrate-and-deploy.yml
Comment (longrunningprocess) [reply .github/workflows/integrate-and-deploy.yml:90]: see above

File: .github/workflows/integrate-and-deploy.yml
Comment (longrunningprocess) [reply .github/workflows/integrate-and-deploy.yml:100]: yes, see above

File: .github/workflows/integrate-and-deploy.yml
Comment (longrunningprocess) [reply .github/workflows/integrate-and-deploy.yml:53]: well, this may not be as straight-forward as I was thinking, it might require looking at a different location to make those things available but we also need to be mindful of the fact that those are incoming workflow args...maybe we should pair on this one.

File: docker/next-app/Dockerfile.next-app
Comment (longrunningprocess) [reply docker/next-app/Dockerfile.next-app:10]: I don't think so...we publish this image and then utilize it in our runtime so everything needs to be in the image.

File: next-app/.npmrc
Comment (longrunningprocess) [reply next-app/.npmrc:1]: yeah, that's probably a good idea, I like the idea of doing that when we move the Makefile to the root of the project, are you ok with giving yourself a reminder for this one?

File: .github/workflows/integrate-and-deploy.yml
Comment (megahirt) [reply .github/workflows/integrate-and-deploy.yml:53]: I defer to your experience and that you tried to make this work.  satisfied for now.

File: .github/workflows/integrate-and-deploy.yml
Comment (megahirt) [reply .github/workflows/integrate-and-deploy.yml:100]: satisfied that you tried to make this work.

File: docker/next-app/Dockerfile.next-app
Comment (megahirt) [reply docker/next-app/Dockerfile.next-app:10]: this is just a comment really.  I think we need to discuss our npm and Docker strategy more so we're on the same page.

File: next-app/.npmrc
Comment (megahirt) [reply next-app/.npmrc:1]: yup, I'm looking forward to the challenge :)

File: next-app/src/lib/forms/Input.svelte
Comment (rmunn) [inline next-app/src/lib/forms/Input.svelte:24]: I see the benefit here for using `type=password` yet having just one Input component instead of two. I think for things like checkboxes and radio buttons we'll want components named Checkbox and Radio, and so on, rather than re-using the Input component.

The one thing I'd want to check is whether this can cause a flash of unstyled content, and more importantly, whether there are any browsers that won't take the `type = 'password'` assignment after DOM rendering, and would therefore cause this to act like a regular text input where the password is visible. I don't *think* there are any such browsers, but I'd like some certainty.

The alternative, where we create a duplicate component called PasswordInput that functions almost identically to an Input component, is code duplication, and I'm fine with this solution **if** no browsers exist where it could expose someone's password to a shoulder-surfer. But I don't know if all mobile browsers (mobile Opera, for instance) would do that.

File: next-app/src/routes/password/changed.svelte
Comment (rmunn) [inline next-app/src/routes/password/changed.svelte:2]: There might be an issue with prerendering pages like this one; the `getSession` hook might not run when the user clicks on links to the rest of the site. See https://github.com/sveltejs/kit/issues/4426. Though that bug will probably be solved by the time we deploy this.

File: next-app/src/lib/fetch/server.js
Comment (rmunn) [inline next-app/src/lib/fetch/server.js:55]: At least for the deployment to QA, I think I'd like this to be a slightly more informative error message, i.e. `catch (e)` and then have `throwError` take an optional third parameter that would be the original exception. We might have code in `throwError` that only passes on that exception if the environment is "dev" or "qa", since we don't want to expose possibly-sensitive exception info on prod. But since I expect that the precise exception thrown would be useful in diagnosing any network errors, I'd like to keep it around in some form unless we're in prod.

File: next-app/src/lib/forms/Button.svelte
Comment (rmunn) [inline next-app/src/lib/forms/Button.svelte:13]: Does the `btn-error` style override `btn-primary`?

Also, would we want `export let primary` in the future, and/or a few other colors-as-props?

**Review by rmunn — APPROVED** [review]
LGTM, with specific points of discussion below. None of them are important enough to hold up merging this to develop, but I'd like to at least have a conversation about these before deploying this to production.

File: next-app/src/lib/progress/index.js
Comment (megahirt) [reply next-app/src/lib/progress/index.js:21]: agreed

File: next-app/src/routes/password/changed.svelte
Comment (megahirt) [reply next-app/src/routes/password/changed.svelte:2]: @rmunn nice job with your Svelte issue repro and getting the attention of the original dev! 💯

File: .github/workflows/build-and-publish-next.yml
Comment (longrunningprocess) [reply .github/workflows/build-and-publish-next.yml:23]: yeah, I'm definitely open to this idea... I'd like to simplify things if possible.

File: .github/workflows/integrate-and-deploy.yml
Comment (longrunningprocess) [reply .github/workflows/integrate-and-deploy.yml:68]: yep, got it

File: next-app/src/lib/progress/index.js
Comment (longrunningprocess) [reply next-app/src/lib/progress/index.js:21]: yeah, I could see a need for that.   Those are called "determinate" progress indicators and we should be able to extend our current "indeterminate" API and component to accept and utilize something like what you've described.

File: next-app/src/lib/forms/Input.svelte
Comment (longrunningprocess) [reply next-app/src/lib/forms/Input.svelte:24]: > I see the benefit here for using type=password yet having just one Input component instead of two. I think for things like checkboxes and radio buttons we'll want components named Checkbox and Radio, and so on, rather than re-using the Input component.

yeah, those might make sense to do for the same reasons we wouldn't include `input type=button` in here but there will be others like `email`, `search` that are similar enough not write a separate comp so it will boil down to a case-by-case basis.  I assume we'll at least do wrappers for the corresponding UI comps from Daisy:
image

> The one thing I'd want to check is whether this can cause a flash of unstyled content

If I'm thinking about the scenario correctly, this would only be a problem if we were to prefill the (password) input with a value which I don't think we ever would.  If there's another scenario I'm not thinking of let me know and we'll REPL it to make sure.

> whether there are any browsers that won't take the type = 'password' assignment after DOM rendering

I'm not familiar with this possibility but if it turns out to be a real concern we can look at mitigating it.

> The alternative, where we create a duplicate component...
yeah, just putting a conditional around the two concrete inputs would be another possible solution.  

We can look at solutions to mitigate your concerns if you find they are still valid.

File: next-app/src/routes/password/changed.svelte
Comment (longrunningprocess) [reply next-app/src/routes/password/changed.svelte:2]: I haven't really put any integrity checks in place on this page to ensure the user came from the correct previous page because if someone is fiddling with url's they can expect some weirdness :-) like a page that says they changed their password when they didn't...so maybe your warning about `getSession` might prove to be relevant for other future pages?

File: next-app/src/lib/fetch/server.js
Comment (longrunningprocess) [reply next-app/src/lib/fetch/server.js:55]: we may need to discuss this more... the exception block you noted here wouldn't get tripped for a `500` from the api call... we wouldn't know about it until we inspected the response.  If it were a 500 we'd need to take a closer look at the shape of the data in that response.  As far as I could tell the only valuable data from `sf` failures was in the fields I pass along...

I did not shut the db down to see what the error response looks like in that case, maybe it's diff.  If so, we'll need to potentially look it various places for the error info.

File: next-app/src/lib/forms/Button.svelte
Comment (longrunningprocess) [reply next-app/src/lib/forms/Button.svelte:13]: > Does the btn-error style override btn-primary?

Yes it does, you can see that demonstrated on the `Index.svelte` page if you just `npm run dev` and hit `http://localhost:3000`

> Also, would we want export let primary in the future, and/or a few other colors-as-props?

Potentially as needed, `secondary` or `accent` perhaps.

File: next-app/src/lib/fetch/server.js
Comment (longrunningprocess) [reply next-app/src/lib/fetch/server.js:55]: fwiw, shutting the db down locally and trying a change password results in the following:
image

File: .github/workflows/build-and-publish-next.yml
Comment (rmunn) [reply .github/workflows/build-and-publish-next.yml:23]: Resolving this conversation so that we won't be blocked from merging the PR. If you have more to say, @longrunningprocess, I'd be happy to carry on the conversation(s) in regular comments.

File: next-app/src/routes/password/changed.svelte
Comment (rmunn) [reply next-app/src/routes/password/changed.svelte:2]: Yes, it probably will... but I hope https://github.com/sveltejs/kit/issues/4426 will be resolved before we need to care about that.

As with the other threads, I'm going to resolve this conversation so it doesn't block our ability to merge the PR. That doesn't mean I don't want to continue discussing this, but it's probably better to do so in primary PR comments so that the PR merge isn't blocked. (Or we could consider whether the "All conversations must be resolved" setting is worth the extra hassle of doing this.)

Comment (longrunningprocess) [reply next-app/src/routes/password/changed.svelte:2]: I still kind of like the resolve rules we have in place.  I'm adding the discussion items to our team meeting agenda

**Comment by megahirt** [comment]
@longrunningprocess I'm good with you merging this.

**Pull request merged**
